bootstrap/config/config.go

```diff
@@ -129,7 +136,11 @@ func ListenForChanges(
 				}
 
 				lc.Info("Writeable configuration has been updated from the Configuration Provider")
-				lc.SetLogLevel(config.GetLogLevel())
+				_ = lc.SetLogLevel(config.GetLogLevel())
+
+				if stream != nil {
+					stream <- struct{}{}
```

Comment:
Send copy of config prior to updating on line 133. This is because we have to detect what changed. See https://github.com/edgexfoundry/app-functions-sdk-go/blob/master/appsdk/sdk.go#L719

Reply:
There's no clean way to do this.  

If I move this to where you want it, there's the potential for the follow-on call to config.UpdateWritableFromRaw() to fail.  Further, you'd be receiving a copy of the current (as yet unmodified) configuration on the stream; there's a potential race condition hidden in there.

If I don't move this, I'd have to copy the underlying structure before calling config.UpdateWritableFromRaw() and subsequently putting the copied struct onto the stream.

IMO, this is a caller's concern; delta detection should be handled there.  You're already likely going to create a goroutine to handle the inbound change notifications.  You can keep your own copy of the configuration struct there to diff with the latest (the same reference that ListenForChanges() is updating).

What do you think @lenny-intel?

Reply:
@michaelestrin Yes, delta detection is caller's concern as long as it has the data to do it. We could grab the initial copy after RunAndReturnWaitGroup() returns. But this will only work for SDKs. If one of the core services decided it need to do something similar, it would have opportunity to grab the initial copy since Run() doesn't return.  

I still think the better solution is to have the copy sent back on the channel.

The race condition can be avoided by simply grabbing a copy before calling config.UpdateWritableFromRaw() and then sending it over the channel where you are currently writing to the channel.  i.e. no moving required. The caller would expect this to be the previous values knowing the current values have already been applied by config.UpdateWritableFromRaw().

Question is then how to you get a copy? I suggest new interface config.CopyOfWritable() which returns an interface{}.

Thoughts?

Reply:
> If one of the core services decided it need to do something similar, it would have opportunity to grab the initial copy since Run() doesn't return.

This is an fence-post edge case and easily solved.  As in your use case, a core service would spin a separate go routine in its service-specific bootstrap handler to handle the updates.  By the time the service-specific bootstrap handler is executed, configuration has already been populated and is available to copy.

> I still think the better solution is to have the copy sent back on the channel.

I don't agree.  You're proposing to add a `CopyOfWritable()`.  This increases complexity and expands the surface area of the contract -- both undesirable.  Further, it would require implementation for concrete instances where it would never be called.

It sounds like you are able to capture your own copy of your service's configuration as part of start up and use that copy to determine what specific properties have changed when a notification is received from the stream.

I believe core services can do the same (notwithstanding that Run() doesn't return until service exit).

Why not take this path?

Reply:
Yep, I think the bootstrap handler will work as long as it has access to the initial config, which we should be able to make happen... I wasn't thinking that the new go routine would be in a bootstrap handler.

Reply:
Doesn't have to be for you.  Does for core. ;)